buildconfig/CMake/CommonSetup.cmake

```diff
@@ -258,20 +258,33 @@ endif ()
 ###########################################################################
 
 if ( CMAKE_VERSION VERSION_GREATER "3.5" )
+  set(DEFAULT_CLANG_TIDY_CHECKS "-*,performance-for-range-copy,performance-unnecessary-copy-initialization,modernize-use-override,modernize-use-nullptr,modernize-loop-convert,modernize-use-bool-literals,modernize-deprecated-headers,misc-*,-misc-unused-parameters")
   set(ENABLE_CLANG_TIDY OFF CACHE BOOL "Add clang-tidy automatically to builds")
   if (ENABLE_CLANG_TIDY)
     find_program (CLANG_TIDY_EXE NAMES "clang-tidy" PATHS /usr/local/opt/llvm/bin )
     if (CLANG_TIDY_EXE)
       message(STATUS "clang-tidy found: ${CLANG_TIDY_EXE}")
-      set(CLANG_TIDY_CHECKS "-*,performance-for-range-copy,performance-unnecessary-copy-initialization,modernize-use-override,modernize-use-nullptr,modernize-loop-convert,modernize-use-bool-literals,modernize-deprecated-headers,misc-*,-misc-unused-parameters")
-      set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_EXE};-checks=${CLANG_TIDY_CHECKS};-header-filter='${CMAKE_SOURCE_DIR}/*'"
-        CACHE STRING "" FORCE)
+      set(CLANG_TIDY_CHECKS "${DEFAULT_CLANG_TIDY_CHECKS}" CACHE STR "Select checks to perform")
+      option(APPLY_CLANG_TIDY_FIX "Apply fixes found through clang-tidy checks" OFF)
+      if(CLANG_TIDY_CHECKS STREQUAL "")
+        # use default checks if empty to avoid errors
+        set(CLANG_TIDY_CHECKS "${DEFAULT_CLANG_TIDY_CHECKS}" CACHE STR "Select checks to perform" FORCE)
+      endif()
+      if(APPLY_CLANG_TIDY_FIX)
+        set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_EXE};-checks=${CLANG_TIDY_CHECKS};-header-filter='${CMAKE_SOURCE_DIR}/*';-fix"
+          CACHE STRING "" FORCE)
+      else()
+        set(CMAKE_CXX_CLANG_TIDY "${CLANG_TIDY_EXE};-checks=${CLANG_TIDY_CHECKS};-header-filter='${CMAKE_SOURCE_DIR}/*'"
+          CACHE STRING "" FORCE)
+      endif()
     else()
       message(AUTHOR_WARNING "clang-tidy not found!")
       set(CMAKE_CXX_CLANG_TIDY "" CACHE STRING "" FORCE) # delete it
     endif()
   else()
     set(CMAKE_CXX_CLANG_TIDY "" CACHE STRING "" FORCE) # delete it
+    set(APPLY_CLANG_TIDY_FIX OFF CACHE BOOL "Apply fixes found through clang-tidy checks" FORCE)
```

Comment:
Is this a common thing in CMake (i.e. unsetting user provided options if they are not valid in the current environment)?

I feel like this should give a warning but leave the users options intact.

Reply:
I'm not sure about CMake in general, but it's been done here as these options don't actually matter if you don't have clang-tidy installed, so we set them to blank to make it clear that you haven't set up clang-tidy

Reply:
Although I suppose `CMAKE_CXX_CLANG_TIDY` is a special cmake parameter so we need to ensure it's blank if we don't have cmake.

I can remove the added lines from 286 if you think it's unnecessary?

Reply:
(I was referring to `APPLY_CLANG_TIDY_FIX` and `CLANG_TIDY_CHECKS`, I agree `CMAKE_CXX_CLANG_TIDY` should be unset)

I'd be more likely to notice a warning rather than variables being unset but if convention supports doing it this way we should do that. @martyngigg?

Reply:
I agree with Dan here. `APPLY_CLANG_TIDY_FIX`/`CLANG_TIDY_CHECKS` are options and  they are passed on a version of cmake <= 3.5 then we're better off warning the developer that they will be ignored here and that they require cmake >= 3.6 to use this feature.

Reply:
p.s. Sorry for not replying sooner - the release prep had my attention.

Reply:
Why is this to do with CMake versions? The check that is failing is for the `clang-tidy` binary being found.

I assume CMake will generate it's own warnings if a version that does not support `option` is used?

Reply:
There are 2 checks here I think. One for cmake >= 3.6. That is the minimum version required to understand `CMAKE_CXX_CLANG_TIDY`. There is then also the check for the `clang-tidy` executable. I guess we need warnings in both.

Reply:
Ah, sorry. I completely missed the >3.5 version check. Yes, both should have warnings.

Reply:
I got confused by the nested ifs there. I think each statement should have the correct error message now